lib/graphs/digraph.nit

```diff
@@ -0,0 +1,147 @@
+# Implementations for representing directed graphs, also called digraphs.
+#
+# Currently, there are two classes
+module digraph
+
+import abstract_digraph
+
+# A directed graph represented by hash maps
+class HashMapDigraph[V, A]
+	super MutableDigraph[V, A]
+
+	# Attributes
+	#
+	private var incoming_arcs_map = new HashMap[V, Array[Arc[V, A]]]
+	private var outgoing_arcs_map = new HashMap[V, Array[Arc[V, A]]]
+	private var number_of_arcs = 0
+
+	redef fun num_vertices: Int do return outgoing_arcs_map.keys.length end
+
+	redef fun num_arcs: Int do return number_of_arcs end
+
+	redef fun add_vertex(u: V)
+	do
+		if not has_vertex(u) then
+			incoming_arcs_map[u] = new Array[Arc[V, A]]
+			outgoing_arcs_map[u] = new Array[Arc[V, A]]
+		end
+	end
+
+	redef fun has_vertex(u: V): Bool do return outgoing_arcs_map.keys.has(u)
+
+	redef fun remove_vertex(u: V)
+	do
+		if has_vertex(u) then
+			for v in successors(u) do
+				remove_arc(u, v)
+			end
+			for v in predecessors(u) do
+				remove_arc(v, u)
+			end
+			incoming_arcs_map.keys.remove(u)
+			outgoing_arcs_map.keys.remove(u)
+		end
+	end
+
+	redef fun add_arc(u, v: V, l: nullable A)
+	do
+		if not has_vertex(u) then add_vertex(u)
+		if not has_vertex(v) then add_vertex(v)
+		if not has_arc(u, v) then
+			var arc = new Arc[V, A](u, v, l)
+			incoming_arcs_map[v].add(arc)
+			outgoing_arcs_map[u].add(arc)
+			number_of_arcs += 1
+		end
+	end
+
+	redef fun has_arc(u, v: V): Bool
+	do
+		if has_vertex(u) and has_vertex(v) then
+			for arc in outgoing_arcs_map[u] do
+				if arc.target == v then return true
+			end
+		end
+		return false
+	end
+
+	redef fun get_arc_value(u, v: V): nullable A do
+		if has_vertex(u) and has_vertex(v) then
+			for arc in outgoing_arcs_map[u] do
+				if arc.target == v then return arc.value
+			end
```

Comment:
I suggest that you add a private `intern_get_arc` that returns the `nullable Arc` that you can use to factorize some code in `has_arc`, `get_arc_value` or `remove_arc`